draft-ietf-privacypass-batched-tokens.md

```diff
@@ -317,15 +326,121 @@ struct {
 ~~~
 
 If the FinalizeBatch function fails, the Client aborts the protocol. Token
-verification works exactly as specified in {{ISSUANCE}}.
+verification works exactly as specified in {{RFC9578}}.
+
+# Arbitrary Batched Token Issuance
+
+This section describes an issuance protocol mechanism for issuing multiple tokens in one round trip between Client and Issuer. An arbitrary batched token request can contain token requests for any token type.
+
+## Client-to-Issuer Request {#arbitrary-client-to-issuer-request}
+
+The Client first creates all TokenRequest it wants to batch. To do so, the client follows
+protocol describing issuance, such as {{RFC9578, Section 5.1}} or {{RFC9578, Section 6.1}}.
+
+The Client then creates a BatchedTokenRequest structured as follows:
+
+~~~tls
+struct {
+  uint16_t token_type;
+  opaque token_request<0..2^16-1>;
+} TokenRequest
+
+struct {
+  TokenRequest token_requests<0..2^16-1>;
+} BatchTokenRequest
+~~~
+
+The structure fields are defined as follows:
+
+- "token_type" is a 2-octet integer, which matches the type of the TokenRequests in the batch.
+
+- "token_requests" are serialized TokenRequests, in network byte order. The number of token_requests, as a 2-octet integer, is prepended to the serialized TokenRequests. In addition, the 2-octet integer length of each TokenRequest is prepended to the serialized TokenRequests. TokenRequest MUST start with a "token_type", not included in the inner opaque token_request attribute.
+
+The Client then generates an HTTP POST request to send to the Issuer Request
+URL, with the BatchTokenRequest as the content. The media type for this request
+is "application/private-token-batch-request". An example request for the Issuer
+Request URL "https://issuer.example.net/request" is shown below.
+
+~~~
+POST /request HTTP/1.1
+Host: issuer.example.net
+Accept: application/private-token-batch-response
+Content-Type: application/private-token-batch-request
+Content-Length: <Length of BatchTokenRequest>
+
+<Bytes containing the BatchTokenRequest>
+~~~
+
+## Issuer-to-Client Response {#arbitrary-issuer-to-client-response}
+
+Upon receipt of the request, the Issuer validates the following conditions:
+
+- The Content-Type is application/private-token-batch-request as registered with IANA.
+
+If this condition is not met, the Issuer MUST return an HTTP 400 error
+to the client.
+
+The Issuer then tries to deserialize the first 4 bytes of the i-th element
+of BatchTokenRequest.token_requests.
+If this is not a token type registered with IANA, the Issuer MUST return an HTTP 400 error
+to the client.
+The issuer creates a BatchTokenResponse structured as follows:
+
+
+~~~tls
+struct {
+  TokenResponse token_responses<0..2^16-1>;
+} BatchTokenResponse
```

Comment:
I think you'll need some syntax changes to support this. How about the following?

```suggestion
struct {
  TokenResponse token_response<0..2^16-1>; /* Defined by token_type */ 
} InnerTokenResponse;

struct {
  InnerTokenResponse token_responses<0..2^16-1>;
} BatchTokenResponse
```

You'd also need to say that a length of 0 for `token_response` indicates it's empty (skipped), where a non-empty length means it's not skipped.

Reply:
i've updated the wording around skip. it was included below, but not clear. I've moved it to the structure definition.

I've used your definition for `InnerTokenRequest`, but without `<0..2^16-1>`, the it is not an array. Maybe we could use opaque again here, but the comment make this clear.